### PR TITLE
feat(nestjs): capture stack traces NEWRELIC-8072

### PR DESCRIPTION
**Pull request by jordigh**
## Proposed Release Notes

* Added support to record Nest.js error stack traces.

## Links

* NEWRELIC-8072

## Details

We instrument `BaseExceptionFilter.handleUnknownError`, which gets called whenever Nest.js throws a 500. This is how Nest.js swallows exceptions, so it seems like a natural hook point.

### CODE REVIEW

**Review by jmartin4563 — CHANGES_REQUESTED** [review]
Couple of comments, I also noticed that when I checkout this branch, that `test/versioned-external/external-repos.js` is different than what's on main (the branches are different), but it isn't showing up here, do you see the same?

File: bin/run-versioned-tests.sh [hunk 1]
```diff
@@ -7,6 +7,7 @@ set -x
 
 VERSIONED_MODE="${VERSIONED_MODE:---minor}"
 SAMPLES="${SAMPLES:-10}"
+export NODE_OPTIONS="--max-old-space-size=4096"
```
[the remark this answers is not in the document]
Comment (jordigh) [reply bin/run-versioned-tests.sh:10]: I just tried this test program:

```sh
#!/bin/bash
LOL=omg
export WTF=brb
env | grep LOL
env | grep WTF
```
and it only outputs `WTF=brb`, as I expected.

So, the `env` binary here starts a new process and as a new process it gets its own environment. The environment of the parent process doesn't all get copied to the child process, only the exported variables.

I'm pretty sure this means that the `export` is required here.

File: lib/shim/webframework-shim.js [hunk 2]
```diff
@@ -36,6 +36,7 @@ const FRAMEWORK_NAMES = {
   HAPI: 'Hapi',
   KOA: 'Koa',
   NEXT: 'Nextjs',
+  NEST: 'Nestjs',
```
[the remark this answers is not in the document]
Comment (jordigh) [reply lib/shim/webframework-shim.js:39]: You're right, I should have called `shim.setFramework(shim.NEST)`, will fix.

**Comment by jordigh** [comment]
> I also noticed that when I checkout this branch, that test/versioned-external/external-repos.js is different than what's on main

That's because my branch isn't based off current main. That change happened after my branch. So in the 3-way merge between current main and my branch, this change gets hidden by the base of the 3-way.

In other words, everything's fine in this regard. :)

File: test/versioned/nestjs/package.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  "name": "nestjs-tests",
+  "version": "0.0.0",
+  "private": true,
+  "tests": [
+    {
+      "engines": {
+        "node": ">=14"
+      },
+      "dependencies": {
+        "@nestjs/cli": ">=8.0.0"
```
Comment (bizob2828) [inline test/versioned/nestjs/package.json:11]: i'll allow @jmartin4563 to finish reviewing but does instrumentation crash if it is less than 8? If so i would do a semver check in the instrumentation

Comment (jordigh) [reply test/versioned/nestjs/package.json:11]: Yeah, I tried 7.0.0 and it failed. `nest build` fails for me with typescript errors, so I'm not sure if it's even supported anymore or not. The least 7.x.x release was in November 2020, so it seems reasonable to drop support for it from our side.

**Review by jmartin4563 — CHANGES_REQUESTED** [review]

File: lib/instrumentation/@nestjs/core.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+'use strict'
+const logger = require('../../logger').child({ component: 'nestjs' })
+const semver = require('semver')
+
+module.exports = function initialize(agent, core, moduleName, shim) {
+  const nestJsVersion = shim.require('./package.json').version
+  shim.setFramework(shim.NEST)
+  // Earliest version that runs in the tests
+  if (semver.lt(nestJsVersion, '8.0.0')) {
+    return
```
[the remark this answers is not in the document]
Comment (jordigh) [reply lib/instrumentation/@nestjs/core.js:18]: Okay done

**Review by jmartin4563 — APPROVED** [review]
thank you!

**Pull request merged**